auth/src/main/java/com/firebase/ui/auth/AuthUI.java

```diff
@@ -401,48 +408,49 @@ public IdpConfig build() {
     }
 
     /**
-     * Builder for the intent to start the user authentication flow.
+     * Base builder for both {@link SignInIntentBuilder} and {@link ReauthIntentBuilder}
      */
-    public final class SignInIntentBuilder {
-        private int mLogo = NO_LOGO;
-        private int mTheme = getDefaultTheme();
-        private LinkedHashSet<IdpConfig> mProviders = new LinkedHashSet<>();
-        private String mTosUrl;
-        private boolean mIsSmartLockEnabled = true;
-        private boolean mAllowNewEmailAccounts = true;
-
-        private SignInIntentBuilder() {
+    @SuppressWarnings(value="unchecked")
+    private abstract class AuthIntentBuilder<T extends AuthIntentBuilder> {
+        int mLogo = NO_LOGO;
+        int mTheme = getDefaultTheme();
+        LinkedHashSet<IdpConfig> mProviders = new LinkedHashSet<>();
+        String mTosUrl;
+        boolean mIsSmartLockEnabled = true;
+
+        private AuthIntentBuilder() {
             mProviders.add(new IdpConfig.Builder(EMAIL_PROVIDER).build());
         }
 
+
         /**
          * Specifies the theme to use for the application flow. If no theme is specified,
          * a default theme will be used.
          */
-        public SignInIntentBuilder setTheme(@StyleRes int theme) {
+        public T setTheme(@StyleRes int theme) {
             Preconditions.checkValidStyle(
-                    mApp.getApplicationContext(),
-                    theme,
-                    "theme identifier is unknown or not a style definition");
+                mApp.getApplicationContext(),
+                theme,
+                "theme identifier is unknown or not a style definition");
             mTheme = theme;
-            return this;
+            return (T) this;
```

Comment:
Sigh.  I looked at this and thought "there's no way that Java really requires this generic + cast situation" but it looks like we do!